Student Name: Manoj Chandra Machavolu<br>
Student Number: 1008518506

In [0]:
#all spark imports
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *

In [0]:
spark = SparkSession.builder.appName("Integer").getOrCreate()

**Part A Q1)**

In [0]:
integerRDD = spark.sparkContext.textFile("/FileStore/tables/integer_1.txt", 8)

In [0]:
integerRDD.collect()

Out[4]: ['45687',
 '5463',
 '34443',
 '223',
 '456',
 '4667',
 '234',
 '9008',
 '1234',
 '2597',
 '6253',
 '10399',
 '3312',
 '2175',
 '20087',
 '6698',
 '31512',
 '618',
 '11698',
 '13969',
 '6575',
 '24846',
 '1313',
 '30977',
 '20927',
 '24771',
 '20293',
 '24210',
 '26005',
 '9680',
 '15297',
 '4099',
 '29625',
 '9374',
 '7645',
 '7660',
 '7025',
 '15755',
 '927',
 '22795',
 '7948',
 '29343',
 '23398',
 '5755',
 '27339',
 '19416',
 '3686',
 '21687',
 '16513',
 '32643',
 '10047',
 '11831',
 '14036',
 '11079',
 '23852',
 '11767',
 '18502',
 '28848',
 '21754',
 '4309',
 '4104',
 '4656',
 '18641',
 '20317',
 '29760',
 '8784',
 '20048',
 '26629',
 '16134',
 '15865',
 '26432',
 '22918',
 '20931',
 '15958',
 '17856',
 '13488',
 '771',
 '28089',
 '15865',
 '22489',
 '5740',
 '12871',
 '28999',
 '9373',
 '31617',
 '8185',
 '19684',
 '25075',
 '17319',
 '20320',
 '21116',
 '32064',
 '15681',
 '11990',
 '12927',
 '31383',
 '2469',
 '21500',
 '4933',
 '21212',
 '12974',
 '367',
 '19773',
 '755

In [0]:
integerRDD.count()

Out[5]: 1010

In [0]:
newRDD = integerRDD.map(lambda x: float(x))

In [0]:
newRDD.collect()

Out[7]: [45687.0,
 5463.0,
 34443.0,
 223.0,
 456.0,
 4667.0,
 234.0,
 9008.0,
 1234.0,
 2597.0,
 6253.0,
 10399.0,
 3312.0,
 2175.0,
 20087.0,
 6698.0,
 31512.0,
 618.0,
 11698.0,
 13969.0,
 6575.0,
 24846.0,
 1313.0,
 30977.0,
 20927.0,
 24771.0,
 20293.0,
 24210.0,
 26005.0,
 9680.0,
 15297.0,
 4099.0,
 29625.0,
 9374.0,
 7645.0,
 7660.0,
 7025.0,
 15755.0,
 927.0,
 22795.0,
 7948.0,
 29343.0,
 23398.0,
 5755.0,
 27339.0,
 19416.0,
 3686.0,
 21687.0,
 16513.0,
 32643.0,
 10047.0,
 11831.0,
 14036.0,
 11079.0,
 23852.0,
 11767.0,
 18502.0,
 28848.0,
 21754.0,
 4309.0,
 4104.0,
 4656.0,
 18641.0,
 20317.0,
 29760.0,
 8784.0,
 20048.0,
 26629.0,
 16134.0,
 15865.0,
 26432.0,
 22918.0,
 20931.0,
 15958.0,
 17856.0,
 13488.0,
 771.0,
 28089.0,
 15865.0,
 22489.0,
 5740.0,
 12871.0,
 28999.0,
 9373.0,
 31617.0,
 8185.0,
 19684.0,
 25075.0,
 17319.0,
 20320.0,
 21116.0,
 32064.0,
 15681.0,
 11990.0,
 12927.0,
 31383.0,
 2469.0,
 21500.0,
 4933.0,
 21212.0,
 12974.0,
 367.0,
 19773.0,
 7553

In [0]:
print('The total number of even numbers = ',newRDD.filter(lambda x: x % 2 == 0).count())

The total number of even numbers =  514


In [0]:
print('The total number of odd numbers = ',newRDD.filter(lambda x: x % 2 != 0).count())

The total number of odd numbers =  496


**Part A Q2)**

In [0]:
salaryRDD = spark.sparkContext.textFile("/FileStore/tables/salary_1.txt", 4)

In [0]:
salaryRDD.collect()

Out[11]: ['Sales 9136',
 'Research 13391',
 'Developer 22220',
 'QA 31888',
 'Marketing 22215',
 'Sales 45567',
 'Research 4023',
 'Developer 7262',
 'QA 5243',
 'Marketing 11425',
 'Sales 11956',
 'Research 24149',
 'Developer 18258',
 'QA 22962',
 'Marketing 28960',
 'Sales 8766',
 'Research 18343',
 'Developer 13686',
 'QA 27626',
 'Marketing 32430',
 'Sales 761',
 'Research 22690',
 'Developer 11703',
 'QA 591',
 'Marketing 15249',
 'Sales 15831',
 'Research 28373',
 'Developer 31034',
 'QA 27821',
 'Marketing 25192',
 'Sales 12640',
 'Research 15100',
 'Developer 2781',
 'QA 17773',
 'Marketing 9047',
 'Sales 20425',
 'Research 19555',
 'Developer 10498',
 'QA 18714',
 'Marketing 1209',
 'Sales 16369',
 'Research 3513',
 'Developer 3362',
 'QA 27723',
 'Marketing 25341',
 'Sales 28399',
 'Research 15762',
 'Developer 23792',
 'QA 13328',
 'Marketing 4390',
 'Sales 32523',
 'Research 16805',
 'Developer 18811',
 'QA 12468',
 'Marketing 7741',
 'Sales 23584',
 'Research 16119',
 'De

In [0]:
salaryMapRDD = salaryRDD.map(lambda x: x.split(" "))
salaryMapRDD.collect()

Out[12]: [['Sales', '9136'],
 ['Research', '13391'],
 ['Developer', '22220'],
 ['QA', '31888'],
 ['Marketing', '22215'],
 ['Sales', '45567'],
 ['Research', '4023'],
 ['Developer', '7262'],
 ['QA', '5243'],
 ['Marketing', '11425'],
 ['Sales', '11956'],
 ['Research', '24149'],
 ['Developer', '18258'],
 ['QA', '22962'],
 ['Marketing', '28960'],
 ['Sales', '8766'],
 ['Research', '18343'],
 ['Developer', '13686'],
 ['QA', '27626'],
 ['Marketing', '32430'],
 ['Sales', '761'],
 ['Research', '22690'],
 ['Developer', '11703'],
 ['QA', '591'],
 ['Marketing', '15249'],
 ['Sales', '15831'],
 ['Research', '28373'],
 ['Developer', '31034'],
 ['QA', '27821'],
 ['Marketing', '25192'],
 ['Sales', '12640'],
 ['Research', '15100'],
 ['Developer', '2781'],
 ['QA', '17773'],
 ['Marketing', '9047'],
 ['Sales', '20425'],
 ['Research', '19555'],
 ['Developer', '10498'],
 ['QA', '18714'],
 ['Marketing', '1209'],
 ['Sales', '16369'],
 ['Research', '3513'],
 ['Developer', '3362'],
 ['QA', '27723'],
 ['Marketing'

In [0]:
print("Salary sum per department")
mappedRDD = salaryMapRDD.map(lambda x: (x[0], int(x[1])))
sumRDD = mappedRDD.reduceByKey(lambda x,y: x+y)
sumRDD.collect()

Salary sum per department
Out[13]: [('Developer', 3221394),
 ('Sales', 3488491),
 ('Research', 3328284),
 ('QA', 3360624),
 ('Marketing', 3158450)]

**Part A Q3)**

In [0]:
textFile = spark.read.text("/FileStore/tables/shakespeare_2.txt")

In [0]:
print('As per the question, only exact words were counted')
print('Shakespeare ',textFile.rdd.map(lambda x: x[0].split(' ').count('Shakespeare')).sum())
print('why ',textFile.rdd.map(lambda x: x[0].split(' ').count('why')).sum())
print('Lord ',textFile.rdd.map(lambda x: x[0].split(' ').count('Lord')).sum())
print('Library ',textFile.rdd.map(lambda x: x[0].split(' ').count('Library')).sum())
print('GLUTENBERG ',textFile.rdd.map(lambda x: x[0].split(' ').count('GUTENBERG')).sum())
print('WILLIAM ',textFile.rdd.map(lambda x: x[0].split(' ').count('WILLIAM')).sum())
print('COLLEGE ',textFile.rdd.map(lambda x: x[0].split(' ').count('COLLEGE')).sum())
print('WORLD ',textFile.rdd.map(lambda x: x[0].split(' ').count('WORLD')).sum())

As per the question, only exact words were counted
Shakespeare  22
why  91
Lord  341
Library  2
GLUTENBERG  99
WILLIAM  115
COLLEGE  98
WORLD  98


**Part A Q4)**

In [0]:
shakespeare_2RDD = spark.sparkContext.textFile("/FileStore/tables/shakespeare_2.txt").flatMap(lambda line: line.split())

In [0]:
mapRDD = shakespeare_2RDD.map(lambda word: (word, 1))
resultRDD = mapRDD.reduceByKey(lambda a,b:a+b)

In [0]:
print('The top 10 most occuring words with their counts are ',resultRDD.map(lambda x: (x[1], x[0])).sortByKey(False).take(11))

The top 10 most occuring words with their counts are  [(11397, 'the'), (8777, 'and'), (8556, 'I'), (7873, 'of'), (7421, 'to'), (5672, 'a'), (4913, 'my'), (4600, 'in'), (4060, 'you'), (3547, 'And'), (3522, 'that')]


In [0]:
print('The top 10 least occuring words with their counts are ',resultRDD.map(lambda x: (x[1], x[0])).sortByKey(True).take(10))

The top 10 least occuring words with their counts are  [(1, 'anyone'), (1, 'restrictions'), (1, 'whatsoever.'), (1, 're-use'), (1, 'online'), (1, 'www.gutenberg.org'), (1, 'COPYRIGHTED'), (1, 'eBook,'), (1, 'Details'), (1, 'guidelines')]


**Part B**

In [0]:
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *

from pyspark.ml.feature import Imputer, StringIndexer, VectorAssembler
from pyspark.ml.linalg import SparseVector, DenseVector
from pyspark.ml import Pipeline
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder

In [0]:
spark = SparkSession.builder.appName("Recommender systems").getOrCreate()

#set the shuffle partition same as number of cpu cores to improve performance 
spark.conf.set("spark.sql.shuffle.partitions", 4)

In [0]:
path = "/FileStore/tables/movies_1.csv"

df = spark.read \
  .format("csv") \
  .option("inferSchema", True) \
  .option("header", True) \
  .option("sep", ',') \
  .option("path", path) \
  .load()

In [0]:
df.display()

movieId rating userId 2 3 0 3 1 0 5 2 0 9 4 0 11 1 0 12 2 0 15 1 0 17 1 0 19 1 0 21 1 0 23 1 0 26 3 0 27 1 0 28 1 0 29 1 0 30 1 0 31 1 0 34 1 0 37 1 0 41 2 0 44 1 0 45 2 0 46 1 0 47 1 0 48 1 0 50 1 0 51 1 0 54 1 0 55 1 0 59 2 0 61 2 0 64 1 0 67 1 0 68 1 0 69 1 0 71 1 0 72 1 0 77 2 0 79 1 0 83 1 0 87 1 0 89 2 0 91 3 0 92 4 0 94 1 0 95 2 0 96 1 0 98 1 0 99 1 0 2 2 1 3 1 1 4 2 1 6 1 1 9 3 1 12 1 1 13 1 1 14 1 1 16 1 1 19 1 1 21 3 1 27 1 1 28 3 1 33 1 1 36 2 1 37 1 1 40 1 1 41 2 1 43 1 1 44 1 1 47 1 1 50 1 1 54 1 1 56 2 1 57 1 1 58 1 1 60 1 1 62 4 1 63 1 1 67 1 1 68 4 1 70 2 1 72 1 1 73 1 1 74 2 1 76 1 1 77 3 1 78 1 1 81 1 1 82 1 1 85 3 1 86 2 1 88 2 1 91 1 1 92 2 1 93 1 1 94 2 1 96 1 1 97 1 1 4 3 2 6 1 2 8 5 2 9 1 2 10 1 2 12 3 2 13 1 2 15 2 2 18 2 2 19 4 2 22 1 2 26 1 2 28 1 2 34 4 2 35 1 2 37 5 2 38 1 2 39 5 2 40 4 2 47 1 2 50 1 2 52 2 2 54 1 2 55 1 2 57 2 2 58 2 2 59 1 2 61 1 2 62 1 2 64 1 2 65 1 2 66 3 2 68 1 2 71 3 2 76 1 2 77 1 2 78 1 2 80 1 2 83 5 2 85 1 2 87 2 2 88 1 2 89 4 2 90 1 2 92 4 2 93 5 2 0 1 3 1 1 3 2 1 3 7 3 3 8 3 3 9 1 3 14 1 3 15 1 3 16 1 3 18 4 3 19 1 3 24 3 3 26 1 3 29 3 3 33 1 3 34 3 3 35 1 3 36 3 3 37 1 3 38 2 3 43 1 3 44 1 3 46 1 3 47 1 3 51 5 3 52 3 3 56 1 3 58 1 3 60 3 3 62 1 3 65 2 3 66 1 3 67 1 3 68 2 3 70 1 3 72 2 3 76 3 3 79 3 3 80 4 3 81 1 3 83 1 3 84 1 3 86 1 3 87 2 3 88 4 3 89 1 3 91 1 3 94 3 3 1 1 4 6 1 4 8 1 4 9 1 4 10 1 4 11 1 4 12 1 4 13 1 4 14 2 4 15 1 4 17 1 4 20 1 4 22 1 4 23 1 4 24 1 4 29 4 4 30 1 4 31 1 4 34 1 4 35 1 4 36 1 4 39 2 4 40 3 4 41 4 4 43 2 4 44 1 4 45 1 4 46 1 4 47 1 4 49 2 4 50 1 4 51 1 4 52 4 4 54 1 4 55 1 4 60 3 4 61 1 4 62 4 4 63 3 4 65 1 4 67 2 4 69 1 4 70 4 4 71 1 4 73 1 4 78 1 4 84 1 4 85 1 4 87 3 4 88 3 4 89 2 4 96 1 4 97 1 4 98 1 4 99 1 4 0 1 5 1 1 5 4 1 5 5 1 5 8 1 5 9 3 5 10 2 5 13 3 5 15 1 5 19 1 5 20 3 5 21 2 5 23 3 5 27 1 5 28 1 5 29 1 5 31 1 5 36 3 5 38 2 5 39 1 5 42 1 5 48 3 5 49 4 5 50 3 5 51 1 5 52 1 5 54 1 5 55 5 5 56 3 5 58 1 5 60 1 5 61 1 5 64 3 5 65 2 5 68 4 5 70 1 5 71 1 5 72 1 5 74 1 5 79 1 5 81 2 5 84 1 5 85 1 5 86 1 5 88 1 5 90 4 5 91 2 5 95 2 5 99 1 5 0 1 6 1 1 6 2 3 6 5 1 6 6 1 6 9 1 6 10 1 6 15 2 6 16 2 6 17 1 6 18 1 6 20 1 6 21 1 6 22 1 6 24 1 6 25 5 6 26 1 6 28 1 6 30 1 6 33 1 6 38 1 6 39 1 6 43 4 6 44 1 6 45 1 6 48 1 6 49 1 6 50 1 6 53 1 6 54 1 6 55 1 6 56 1 6 58 4 6 59 1 6 60 1 6 61 3 6 63 3 6 66 1 6 67 3 6 68 1 6 69 1 6 71 2 6 73 1 6 75 1 6 77 1 6 79 1 6 81 1 6 84 1 6 85 3 6 86 1 6 87 1 6 88 1 6 89 1 6 91 2 6 94 1 6 95 2 6 96 1 6 1 1 7 2 2 7 3 1 7 4 1 7 7 1 7 10 1 7 11 2 7 14 2 7 15 1 7 16 1 7 18 1 7 21 1 7 22 1 7 23 1 7 25 5 7 26 1 7 29 4 7 30 1 7 31 3 7 32 1 7 33 1 7 35 1 7 37 2 7 39 3 7 40 2 7 42 2 7 44 1 7 45 2 7 47 4 7 48 1 7 49 1 7 53 1 7 54 1 7 55 1 7 56 1 7 59 1 7 61 2 7 62 3 7 63 2 7 66 1 7 67 3 7 74 1 7 75 1 7 76 3 7 77 1 7 81 1 7 82 1 7 84 2 7 85 4 7 86 1 7 92 2 7 96 1 7 97 1 7 98 1 7 0 1 8 2 4 8 3 2 8 4 2 8 5 1 8 7 1 8 9 1 8 11 1 8 15 1 8 18 1 8 19 1 8 21 1 8 29 5 8 31 3 8 33 1 8 35 1 8 36 1 8 40 2 8 44 1 8 45 1 8 50 1 8 51 1 8 52 5 8 53 5 8 54 1 8 55 1 8 56 1 8 58 4 8 60 3 8 62 4 8 64 1 8 67 3 8 69 1 8 71 1 8 72 3 8 77 3 8 78 1 8 79 1 8 83 1 8 85 5 8 86 1 8 88 1 8 90 1 8 92 2 8 95 4 8 96 3 8 97 1 8 98 1 8 99 1 8 2 3 9 3 1 9 4 1 9 5 1 9 6 1 9 7 5 9 9 1 9 12 1 9 14 3 9 15 1 9 19 1 9 21 1 9 22 1 9 24 1 9 25 1 9 26 1 9 30 3 9 32 4 9 35 2 9 36 2 9 37 2 9 38 1 9 39 1 9 43 3 9 49 5 9 50 3 9 53 1 9 54 1 9 58 1 9 59 1 9 60 1 9 61 1 9 63 3 9 64 3 9 68 1 9 69 1 9 70 3 9 71 1 9 73 2 9 75 1 9 77 2 9 81 2 9 82 1 9 83 1 9 84 1 9 86 1 9 87 4 9 88 1 9 90 3 9 94 2 9 95 3 9 97 2 9 98 1 9 0 3 10 2 4 10 4 3 10 7 1 10 8 1 10 10 1 10 13 2 10 14 1 10 16 2 10 17 1 10 18 1 10 21 1 10 22 1 10 24 1 10 25 3 10 28 1 10 35 1 10 36 1 10 37 1 10 38 1 10 39 1 10 40 4 10 41 2 10 42 3 10 43 1 10 49 3 10 50 1 10 51 1 10 52 1 10 55 2 10 56 1 10 58 1 10 63 1 10 66 1 10 67 2 10 68 1 10 75 1 10 77 1 10 79 1 10 86 1 10 89 3 10 90 1 10 97 1 10 98 1 10 0 1 11 6 2 11 9 1 11 10 1 11 11 1 11 12 1 11 13 4 11 16 1 11 18 5 11 19 4 11 20 1 11 21 1 11 22 1 11 23 5 11 25 1 11 27 5 

The csv file consists of 3 columns namely movieId, rating and userId.<br>The ratings are out of 5 (5 means very good movie and 1 means poor).<br>

**Part B Q1)**

In [0]:
print("Top 20 movies with highest ratings")
new_df = df.groupBy("movieId").agg(count("rating")).orderBy(col("count(rating)").desc())
new_df.show(20)

Top 20 movies with highest ratings
+-------+-------------+
|movieId|count(rating)|
+-------+-------------+
|     51|           20|
|     50|           20|
|      6|           20|
|     29|           20|
|     22|           20|
|     15|           19|
|     55|           19|
|     68|           19|
|      2|           19|
|     94|           19|
|     86|           18|
|     45|           18|
|     36|           18|
|     14|           18|
|     88|           18|
|     85|           18|
|     19|           17|
|     12|           17|
|     67|           17|
|     56|           17|
+-------+-------------+
only showing top 20 rows



In [0]:
print("Top 15 users who provided the highest ratings")
new_df_2 = df.groupBy("userId").agg(count("rating")).orderBy(col("count(rating)").desc())
new_df_2.show(15)

Top 15 users who provided the highest ratings
+------+-------------+
|userId|count(rating)|
+------+-------------+
|    14|           57|
|     6|           57|
|    22|           56|
|    11|           56|
|    12|           55|
|     4|           55|
|     7|           54|
|     9|           53|
|    23|           52|
|    24|           52|
|    18|           52|
|    28|           50|
|    19|           49|
|     1|           49|
|     5|           49|
+------+-------------+
only showing top 15 rows



**Part B Q2)**

In [0]:
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.recommendation import ALS
from pyspark.ml.tuning import ParamGridBuilder, TrainValidationSplit, CrossValidator

Combination 1: 60% Train set and 40% Test set

In [0]:
(training_1, test_1) = df.randomSplit([0.6, 0.4], 20)   #### Setting a constant seed value so that the split remais constant throughout

In [0]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als_1 = ALS(userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model_1 = als_1.fit(training_1)

In [0]:
# Evaluate the model by computing the RMSE on the test data
predictions_1 = model_1.transform(test_1)
predictions_1.display()

movieId rating userId prediction 0 1 13 0.91065717 3 1 13 0.86477196 4 3 18 1.2804631 5 1 13 1.1197567 6 1 13 1.0662738 7 1 18 -0.039926007 7 3 12 2.5876398 9 1 14 1.0764339 10 1 14 0.8035929 11 1 14 0.83973676 11 4 18 0.54803 12 2 18 1.8793347 12 3 25 0.92887706 13 1 14 1.25333 14 3 14 1.4606191 15 2 12 0.90006906 16 1 14 1.1545506 16 4 12 1.1312804 17 1 13 1.0633267 17 1 14 1.1519139 18 1 25 1.1217966 18 3 13 1.6381446 20 1 13 0.686671 20 1 18 0.75530076 21 1 12 1.0292208 21 1 13 1.1411813 21 1 14 1.0224056 22 1 13 1.4452878 22 1 18 0.23951636 22 1 25 1.0979025 23 1 25 2.1951559 25 1 12 1.327767 25 1 18 1.2476052 26 1 18 0.65905684 26 2 25 0.7318922 27 1 13 1.1327828 27 1 14 1.6382226 28 1 14 0.6753956 29 5 14 3.8427303 30 1 25 1.0972507 30 2 12 1.5146883 32 1 18 1.8123837 33 4 25 1.5095215 35 5 12 2.1489701 37 1 25 1.087681 37 2 14 1.8826385 39 5 18 0.66734344 40 1 14 1.7177585 40 1 18 0.601729 40 2 13 2.167268 41 1 25 0.9797187 44 4 18 0.09520388 45 1 13 1.1760713 47 1 18 1.2733034 48 1 18 1.7392917 49 1 13 1.4312005 51 1 14 2.2713785 52 1 12 1.1121774 53 3 14 1.3601289 54 1 13 1.1104974 56 1 14 1.1858133 56 1 18 0.8157492 56 2 12 1.628561 58 2 25 1.1244557 62 1 13 2.4310951 63 1 12 1.0298012 63 5 14 3.3175786 64 1 13 0.5681574 66 3 18 0.004733324 68 1 13 1.1845458 68 1 14 1.6988674 68 1 25 1.0939896 70 1 18 0.8284646 70 1 25 0.8940577 72 1 12 1.9159267 72 3 13 1.4265112 73 1 13 0.69672316 73 1 25 1.1595045 74 1 12 1.6250607 74 3 13 2.1336386 75 1 12 2.0571573 76 1 25 0.9329288 76 3 18 0.5731387 77 1 18 0.7341764 77 2 13 1.7719593 78 1 13 0.84804237 79 3 12 1.9661176 80 3 18 0.79984474 81 3 18 1.9231108 82 1 14 1.1053429 82 1 18 1.2425761 83 1 12 1.075937 83 3 13 1.6885126 83 5 18 0.81111985 84 1 12 0.6198063 85 1 12 1.0919566 86 1 13 0.7145467 90 1 13 1.0928535 91 3 12 1.7115384 91 3 25 1.523038 92 1 25 2.2842743 93 4 13 0.8844278 94 1 13 1.2611512 94 1 14 1.35923 94 1 25 0.6301102 95 2 14 2.5133867 99 1 13 0.786134 99 2 18 0.7367119 0 1 6 0.84229934 0 1 23 1.1073289 1 1 6 1.6077886 1 1 19 1.224354 1 1 20 1.6788937 2 1 15 1.5272818 2 1 23 1.1395488 2 3 6 1.9956367 4 1 19 0.68466395 4 1 23 0.888055 5 1 29 1.3116829 5 3 16 1.3731476 6 1 9 0.8682511 6 1 15 0.935156 6 1 20 1.5853438 6 2 19 1.0364016 6 2 23 1.3272305 7 1 15 1.7077305 9 1 6 1.3114748 10 1 6 1.1712706 11 1 29 0.7647472 12 1 19 0.4823897 13 3 29 2.3748565 14 1 19 1.4959359 14 1 23 1.4854723 14 1 29 1.1130674 15 1 16 0.73458517 17 5 17 3.1397424 19 1 16 0.69347394 19 3 29 1.8295422 21 2 20 2.0586026 22 1 9 2.8248167 22 2 15 1.6335677 22 3 29 2.2155263 23 2 15 1.8122209 23 4 23 4.2212486 25 1 15 1.1883526 25 1 20 1.0242609 25 5 6 1.6392471 26 1 20 0.9504568 26 1 23 0.9225446 26 3 15 0.6837687 27 1 15 1.6494269 29 1 15 0.93378764 29 2 20 2.6103883 29 4 16 2.0642548 30 1 6 1.2345014 30 2 20 2.0159957 30 4 23 2.1972008 31 1 29 1.8487301 32 4 9 2.5266519 32 4 29 2.7019815 33 1 6 1.1500381 33 2 23 1.6700513 34 1 17 0.95586723 34 1 19 0.28549847 35 2 15 1.7253116 36 2 29 2.2063487 37 1 17 1.256319 37 1 23 2.0500565 38 1 6 0.4787219 38 1 9 0.10237 38 1 19 1.5531085 38 1 23 2.475385 38 3 29 2.081126 39 1 6 2.3015647 39 1 9 1.6735779 39 1 23 0.848011 39 1 29 0.97922856 40 1 20 1.1536014 41 1 19 1.6382421 43 1 19 1.5557292 43 1 23 2.2801988 43 4 6 1.1578388 45 1 6 1.1657227 46 5 15 0.6792573 46 5 17 0.5955276 46 5 29 0.65830624 47 1 17 1.6671232 49 3 29 2.64366 49 5 23 3.4444578 50 2 15 1.3271129 50 3 9 2.306682 50 4 23 4.1668005 51 1 17 2.373109 51 2 29 1.7947204 53 1 23 0.839186 53 2 19 2.3006725 53 3 20 1.6639729 54 1 9 1.4011403 54 5 16 1.8420951 55 1 16 0.6554277 55 5 17 3.140035 58 1 9 1.645695 58 1 15 0.7790897 58 4 6 1.3320466 59 1 9 1.7353833 59 1 17 1.199846 60 1 6 1.2861444 60 1 16 1.5275956 60 1 17 1.025851 60 1 20 2.2172205 61 1 9 1.3354739 61 1 20 1.2861454 61 1 23 1.150059 62 1 15 0.94595194 62 1 19 3.002438 62 1 29 1.4649237 64 1 20 1.1978204 64 3 15 1.7814951 64 4 23 3.31007 65 1 19 0.9608906 65 2 15 0.6781729 65 5 23 1.2488964 66 1 17 0.94179183 68 1 9 3.42

Combination 2: 80% Train set and 20% Test set

In [0]:
(training_2, test_2) = df.randomSplit([0.8, 0.2], 20)   #### Setting a constant seed value so that the split remais constant throughout

In [0]:
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als_2 = ALS(userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model_2 = als_2.fit(training_2)

In [0]:
    predictions_2 = model_2.transform(test_2)
predictions_2.display()

movieId rating userId prediction 12 2 18 1.9533653 13 1 14 0.8958125 14 3 14 2.261268 18 3 13 1.6363255 25 1 12 0.86819506 25 1 18 1.3692902 37 1 25 1.3540595 56 1 14 1.0286055 56 1 18 0.81377834 56 2 12 2.0548666 70 1 25 1.1626883 73 1 13 0.80928934 83 1 12 1.6702003 83 3 13 1.656797 93 4 13 2.054863 12 1 19 0.23681518 13 3 29 2.6998107 14 1 23 1.4718714 14 1 29 1.1028072 25 1 15 1.3135507 25 5 6 1.0129094 38 1 6 -0.29604352 38 1 9 0.27188343 46 5 15 1.8463883 46 5 17 3.8375077 50 3 9 2.3263288 50 4 23 3.6544104 65 1 19 1.3412931 65 5 23 2.3058753 73 1 15 1.3975602 73 2 20 1.2514038 83 1 29 1.754024 95 2 6 2.0234728 97 1 15 0.84936714 12 1 4 1.2732583 12 3 2 2.853737 13 1 2 1.7986771 13 2 10 1.4279433 25 1 22 1.1099124 56 1 10 1.0582376 67 2 10 1.3296905 70 1 5 1.5077438 95 2 5 1.7460026 12 1 11 1.8686699 18 4 3 2.9965832 18 4 27 2.363534 25 5 7 1.5162512 37 1 3 1.0415941 37 1 24 2.1285443 46 1 21 0.7233211 50 1 24 2.0849864 50 2 26 1.865142 56 2 1 1.6660333 65 1 24 1.389017 67 1 3 0.76327693 67 1 24 1.9446971 67 3 8 2.8276906 73 4 26 1.5240277 83 1 8 0.87662196 83 1 21 1.3763124 83 3 27 1.5715271 93 1 1 1.2053763 93 1 27 1.0811986 97 1 1 0.7788492 6 1 13 1.1040045 16 1 14 0.6845912 20 1 13 0.66291314 20 1 18 0.47540525 30 2 12 3.396617 39 5 18 1.7867496 40 1 14 1.6756864 40 1 18 2.1522114 44 4 18 1.367809 48 1 18 1.6249655 51 1 14 2.1094556 62 1 13 2.2162628 68 1 13 1.2846751 68 1 14 1.51245 74 1 12 1.5577017 84 1 12 1.3773314 94 1 13 1.0701722 99 1 13 0.8342224 6 1 9 1.0329014 6 1 20 1.5734338 6 2 19 1.6416457 6 2 23 1.2730184 17 5 17 2.9476762 19 1 16 0.60688496 23 2 15 1.1434098 23 4 23 4.0566206 30 1 6 2.00567 30 2 20 1.144378 39 1 6 2.7964654 39 1 9 2.32606 43 1 23 1.7774929 51 1 17 1.9577013 51 2 29 1.46778 53 3 20 1.645117 54 1 9 1.0793179 54 5 16 1.8468523 61 1 9 1.9791741 61 1 23 1.4310293 62 1 15 1.5920718 64 1 20 0.73033726 64 3 15 2.0146441 64 4 23 3.2253003 68 1 9 2.655169 74 1 15 1.7438002 74 1 29 1.251496 81 1 6 0.84265524 81 1 19 0.2358386 82 1 16 1.2136874 82 1 19 0.88950956 82 2 23 1.6520237 84 1 9 0.9731134 94 1 6 1.1241784 94 4 17 3.7682886 94 4 20 2.6795454 94 4 29 2.6359956 6 1 2 1.326034 9 1 4 1.4931442 9 1 22 1.721661 9 3 5 1.2832799 16 1 22 1.2723185 16 2 10 1.10516 17 1 10 0.7486624 20 3 5 3.1667194 23 3 28 2.2803814 30 5 22 1.6718239 42 1 5 1.1976281 43 1 10 0.94181705 44 1 4 0.9627254 44 1 28 1.6015472 51 5 22 3.3184526 54 1 2 1.2029153 61 1 2 2.4777334 64 3 5 1.9878492 81 1 22 0.7802906 86 1 5 1.0447978 9 1 8 1.2921726 9 1 11 0.40396374 9 4 0 0.78479993 16 1 3 0.7005336 17 1 0 1.0067267 19 1 1 0.805102 19 1 3 1.5253624 19 1 24 1.794486 19 3 27 2.1024861 20 1 21 0.7502744 20 1 26 2.0080972 23 1 7 0.28983182 29 1 27 1.215996 30 1 7 1.6357439 30 2 21 2.207294 30 5 11 4.2088842 39 1 24 1.9062599 40 1 1 1.5830497 40 1 24 0.9259232 43 3 21 2.4515018 43 3 24 2.6721146 48 1 26 2.0179815 53 5 8 2.890129 54 1 0 1.1768428 54 1 1 1.663637 54 1 7 0.66498893 54 1 8 0.8849536 61 1 27 0.79288363 62 1 3 2.6977162 68 4 1 2.216029 68 4 26 3.1859488 71 1 8 1.425055 71 1 21 1.0717779 71 3 11 1.9919621 81 3 26 1.7748737 84 2 7 1.2062155 86 1 7 1.0290183 86 1 8 1.1514865 86 1 11 0.9219683 86 3 26 1.0833335 94 2 1 1.9238869 94 2 27 1.737466 94 5 26 3.1992426 22 1 13 1.3231715 28 1 14 1.5878623 45 1 13 0.9525724 76 3 18 0.7337708 77 1 18 1.1296303 77 2 13 1.8989683 92 1 25 1.5004065 2 1 15 1.4749212 2 3 6 1.9857575 4 1 19 1.4869982 4 1 23 0.9133426 5 3 16 1.1516244 22 2 15 2.0026023 22 3 29 1.7696105 32 4 9 3.6006148 49 3 29 3.3979986 55 1 16 1.5621277 58 1 9 2.4567776 59 1 9 1.2471789 77 1 6 1.6115499 79 1 6 0.8928344 88 1 23 1.1077461 90 3 20 2.4249325 2 4 10 2.7217627 5 1 5 0.9780334 10 1 10 1.2623069 28 1 2 2.879273 28 1 10 1.9299232 31 1 5 1.0784945 36 3 22 2.761569 49 2 4 1.8778316 59 1 2 0.8816887 60 1 28 1.0492655 88 2 28 1.1340628 88 3 4 1.7357213 2 3 0 2.1694725 2 4 21 2.647888 10 2 24 1.2458702 28 1 27 2.2897735 31 1 24 2.6811643 31 1 26 -0.023151338 45 1 21 1.1154606 45 2 26 1.8886302 49 1 7 1.383815 55 1 8 0

**Part B Q3)**

**MAE - Mean Absolute Error**<br>
MAE evaluates the absolute distance of the observations (the entries of the dataset) to the predictions on a regression, taking the average over all observations. We use the absolute value of the distances so that negative errors are accounted properly.<br>
Formula for MAE - https://miro.medium.com/max/510/1*V5kDRZyG6udD7tKStFotlQ.png<br>
**MSE - Mean Squared Error**<br>
MSE squares the distance between the actual and prediction values, taking the average over all observations. Higher errors (or distances) weigh more in the metric than lower ones, due to the nature of the power function. A backlash in MSE is the fact that the unit of the metric is also squared, so if the model tries to predict price in US$, the MSE will yield a number with unit (US$)² which does not make sense.<br>
Formula for MSE - https://miro.medium.com/max/532/1*VZ3iKUrzH7f71B03Sui6fQ.png<br>
**RMSE - Root Mean Squared Error**<br>
RMSE is used to return the MSE error to the original unit by taking the square root of it, while maintaining the property of penalizing higher errors.<br>
Formula for RMSE -  https://miro.medium.com/max/640/1*YDOgUMYYjETdAVAn7xEOrw.png

**MAE and RMSE — Which Metric is Better?**  Link - https://medium.com/human-in-a-machine-world/mae-and-rmse-which-metric-is-better-e60ac3bde13d<br>RMSE has the benefit of penalizing large errors more so can be more appropriate in some cases.<br>Thus we prefer RMSE over MAE and MSE

In [0]:
# Evaluate the model by computing the RMSE on the test data 1
evaluator_1 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse_1 = evaluator_1.evaluate(predictions_1)
print("Root-mean-square error = " + str(rmse_1))

Root-mean-square error = 1.1427514549626354


In [0]:
# Evaluate the model by computing the RMSE on the test data 2
evaluator_2 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse_2 = evaluator_2.evaluate(predictions_2)
print("Root-mean-square error = " + str(rmse_2))

Root-mean-square error = 1.0365173812132722


Thus, the dataset with train test split of 80-20 is effective and has less errors since a larger part of data (80%) is used for training the model compared to only 60%.

**Part B Q4)**

Hyperparameter tuning of model 1

In [0]:
paramGrid = ParamGridBuilder()\
    .addGrid(als_1.regParam, [0.1, 0.01]) \
    .addGrid(als_1.maxIter, [3,4,5])\
    .addGrid(als_1.alpha, [0.4,0.8])\
    .build()

In [0]:
crossval_1 = CrossValidator(estimator=als_1,
                           estimatorParamMaps=paramGrid,
                           evaluator=evaluator_1,
                           numFolds=3)

In [0]:
model_1 = crossval_1.fit(training_1)

In [0]:
model_1_prediction = model_1.transform(test_1)
rmse_1_tuned = evaluator_1.evaluate(model_1_prediction)
print("Root-mean-square error = " + str(rmse_1_tuned))

Root-mean-square error = 1.2186136086995472


In [0]:
import numpy as np
print(model_1.getEstimatorParamMaps()[np.argmax(model_1.avgMetrics)])

{Param(parent='ALS_2a02acedeada', name='regParam', doc='regularization parameter (>= 0).'): 0.01, Param(parent='ALS_2a02acedeada', name='maxIter', doc='max number of iterations (>= 0).'): 3, Param(parent='ALS_2a02acedeada', name='alpha', doc='alpha for implicit preference'): 0.4}


Thus the best parameters for model 1 (60/40 split) are : regParam = 0.01, maxIter = 3 and alpha = 0.4<br>The RMSE of model 1 after tuning is 1.2186

Hyperparameter tuning of model 2

In [0]:
crossval_2 = CrossValidator(estimator=als_2,
                           estimatorParamMaps=paramGrid,
                           evaluator=evaluator_2,
                           numFolds=3)

In [0]:
model_2 = crossval_2.fit(training_2)

In [0]:
model_2_prediction = model_2.transform(test_2)
rmse_2_tuned = evaluator_2.evaluate(model_2_prediction)
print("Root-mean-square error = " + str(rmse_2_tuned))

Root-mean-square error = 1.0365173812132722


In [0]:
print(model_2.getEstimatorParamMaps()[np.argmax(model_2.avgMetrics)])

{Param(parent='ALS_2a02acedeada', name='regParam', doc='regularization parameter (>= 0).'): 0.1, Param(parent='ALS_2a02acedeada', name='maxIter', doc='max number of iterations (>= 0).'): 3, Param(parent='ALS_2a02acedeada', name='alpha', doc='alpha for implicit preference'): 0.8}


Thus the best parameters for model 2 (80/20 split) are regParam = 0.1, maxIter = 4 and alpha = 0.8<br>The RMSE of model 2 after tuning is 1.0365 which is same as before tuning.<br> Hence my recommendation is to not tune the hyper parameters as it just leads to increase in computation time.

**Comparing the 2 models after Hyper parameter tuning, the best model is model 2 with 80/20 split ratio of RMSE 1.0365**<br>
More the data used for training, better is the prediction

**Significance of parameters used for Hyper tuning.**<br>
**maxIter** is the maximum number of iterations to run (defaults to 10).<br>
**regParam** specifies the regularization parameter in ALS (defaults to 1.0).<br>
**alpha** is a parameter applicable to the implicit feedback variant of ALS that governs the baseline confidence in preference observations (defaults to 1.0).

**Part 2 Q5)**

Now, extracting the best hyper paramters and using model 2 in ALS, we get

In [0]:
als = ALS(maxIter=4, regParam=0.1, alpha=0.8, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")
model = als.fit(training_2)

In [0]:
predictions = model.transform(test_2)

In [0]:
userRecs = model.recommendForAllUsers(15)
userRecs.display()

/databricks/spark/python/pyspark/sql/context.py:134: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(


userId recommendations 20 List(List(22, 3.5436418), List(77, 3.1543994), List(88, 2.9691494), List(62, 2.691823), List(51, 2.6850338), List(74, 2.6340606), List(7, 2.4530618), List(98, 2.38503), List(54, 2.3844337), List(94, 2.3465044), List(24, 2.293388), List(32, 2.1524203), List(92, 2.1471808), List(87, 2.1471012), List(41, 2.0527668)) 10 List(List(49, 2.9692636), List(40, 2.7998114), List(92, 2.5178447), List(42, 2.4445102), List(47, 2.202442), List(4, 2.197721), List(25, 2.1546082), List(0, 2.0897472), List(2, 2.0821645), List(81, 2.0391595), List(85, 1.9937572), List(89, 1.960944), List(1, 1.8860781), List(93, 1.7398272), List(12, 1.7333003)) 0 List(List(92, 2.8872342), List(62, 2.2493389), List(2, 2.1866317), List(49, 1.9753945), List(89, 1.9657198), List(1, 1.9413546), List(12, 1.9109589), List(4, 1.826584), List(22, 1.8096465), List(77, 1.7510881), List(95, 1.7288542), List(91, 1.7239553), List(41, 1.7123939), List(57, 1.6924877), List(81, 1.6646932)) 1 List(List(62, 2.7366626), List(77, 2.6835294), List(98, 2.559925), List(29, 2.497772), List(90, 2.429508), List(28, 2.414606), List(85, 2.4139247), List(94, 2.3954282), List(22, 2.3829741), List(52, 2.3825567), List(74, 2.297678), List(51, 2.2233999), List(92, 2.2078123), List(41, 2.1721256), List(88, 2.0996337)) 11 List(List(27, 4.6242313), List(23, 4.527977), List(32, 4.514224), List(69, 4.504795), List(48, 4.256778), List(18, 4.1610374), List(79, 4.1162205), List(90, 4.075428), List(34, 3.770828), List(66, 3.6660514), List(19, 3.6472616), List(13, 3.576232), List(81, 3.5759928), List(38, 3.4879177), List(30, 3.3979263)) 21 List(List(29, 4.41662), List(52, 4.207973), List(53, 3.8503525), List(62, 3.808525), List(74, 3.487352), List(76, 3.437942), List(41, 3.073553), List(70, 2.8862903), List(63, 2.8537436), List(85, 2.8116713), List(58, 2.7631187), List(72, 2.720936), List(93, 2.7039273), List(96, 2.5343096), List(67, 2.5243573)) 12 List(List(27, 4.7341595), List(32, 4.4351935), List(35, 4.3454323), List(17, 4.1923122), List(55, 4.000731), List(48, 3.9757478), List(64, 3.9035752), List(30, 3.903024), List(49, 3.8981318), List(90, 3.8806548), List(46, 3.8727002), List(50, 3.5768325), List(16, 3.539682), List(94, 3.4145155), List(31, 3.0832968)) 2 List(List(83, 4.499529), List(93, 4.4600716), List(39, 4.053454), List(37, 3.9827049), List(92, 3.8548613), List(40, 3.5830336), List(89, 3.5568874), List(34, 3.492097), List(19, 3.4342837), List(81, 3.2545238), List(4, 3.0386217), List(49, 3.0089474), List(28, 3.001029), List(23, 2.9694078), List(43, 2.9105105)) 22 List(List(22, 4.464737), List(74, 4.428699), List(88, 4.31415), List(77, 3.6875594), List(32, 3.5696619), List(62, 3.5036075), List(7, 3.460813), List(51, 3.3258634), List(98, 3.2181628), List(54, 3.213729), List(24, 3.1487114), List(41, 3.025909), List(53, 2.9823115), List(29, 2.977372), List(94, 2.963779)) 13 List(List(74, 2.774088), List(29, 2.5024345), List(41, 2.463862), List(53, 2.4456398), List(52, 2.416337), List(62, 2.4118762), List(93, 2.3580184), List(92, 2.2555149), List(70, 2.1916344), List(28, 2.0996432), List(7, 2.0223842), List(39, 2.006426), List(87, 1.9602845), List(37, 1.9282062), List(83, 1.9176053)) 23 List(List(32, 4.8268714), List(55, 4.695011), List(90, 4.5592794), List(49, 4.3759985), List(48, 4.2517295), List(27, 4.1733418), List(23, 4.0698647), List(46, 4.062708), List(20, 3.8487244), List(30, 3.8131819), List(17, 3.7686944), List(13, 3.500413), List(35, 3.500189), List(94, 3.4959552), List(50, 3.4611094)) 3 List(List(51, 4.230642), List(22, 3.7156234), List(88, 3.6819751), List(32, 3.5737932), List(23, 3.3978038), List(90, 3.335126), List(18, 3.3037364), List(80, 3.3014119), List(74, 3.1609747), List(62, 2.9615424), List(24, 2.9361389), List(69, 2.8810246), List(77, 2.8563063), List(29, 2.8362885), List(76, 2.803315)) 14 List(List(29, 4.7187376), List(52, 4.655295), List(76, 4.1710854), List(85, 4.0850472), List(62, 3.7639232), List(96, 3.279928), List(63, 3.1719012), List(72, 3.1679373

In [0]:
import pyspark.sql.functions as f
userRecs.filter(f.col('userId')==10).display()

userId recommendations 10 List(List(49, 2.9692636), List(40, 2.7998114), List(92, 2.5178447), List(42, 2.4445102), List(47, 2.202442), List(4, 2.197721), List(25, 2.1546082), List(0, 2.0897472), List(2, 2.0821645), List(81, 2.0391595), List(85, 1.9937572), List(89, 1.960944), List(1, 1.8860781), List(93, 1.7398272), List(12, 1.7333003))

In [0]:
userRecs.filter(f.col('userId')==14).display()

userId recommendations 14 List(List(29, 4.7187376), List(52, 4.655295), List(76, 4.1710854), List(85, 4.0850472), List(62, 3.7639232), List(96, 3.279928), List(63, 3.1719012), List(72, 3.1679373), List(69, 2.9123392), List(53, 2.908828), List(90, 2.886309), List(47, 2.8856504), List(30, 2.8075178), List(58, 2.7764516), List(67, 2.7576265))

There is a chance that the movie recommended to a user is already watched and rated by that user.<br> Hence we need to exclude the movieId already rated by the user in order to get new recommendations.

In [0]:
userData = df.filter(df.userId.isin([10,14]))
userSubsetRecs = model.recommendForUserSubset(userData, 100)

In [0]:
newRecommendationsDF = (userSubsetRecs.select('userId',explode('recommendations').alias('recommendation')).select('userId','recommendation.*')).sort('rating',ascending=False)
newRecommendationsDF.select('movieId','rating','userId')

Out[50]: DataFrame[movieId: int, rating: float, userId: int]

In [0]:
print("Top 15 movies recommended to user 10 are ")
result1 = newRecommendationsDF.join(userData, on =['movieId'],how ='left_anti')
result1.filter(result1.userId.isin([10])).show(15)

Top 15 movies recommended to user 10 are 
+-------+------+----------+
|movieId|userId|    rating|
+-------+------+----------+
|     92|    10| 2.5178447|
|     81|    10| 2.0391595|
|     85|    10| 1.9937572|
|     91|    10| 1.6262096|
|     87|    10| 1.5756335|
|     57|    10| 1.5597557|
|     19|    10| 1.3917103|
|     33|    10|  1.305766|
|     61|    10| 1.2916336|
|     65|    10| 1.1958983|
|     70|    10| 1.1757578|
|     23|    10| 0.9971097|
|     30|    10|0.98390806|
|     99|    10| 0.9517034|
|     64|    10| 0.9479872|
+-------+------+----------+
only showing top 15 rows



In [0]:
print("Top 15 movies recommended to user 14 are ")
result2 = newRecommendationsDF.join(userData, on =['movieId'],how ='left_anti')
result2.filter(result2.userId.isin([14])).show(15)

Top 15 movies recommended to user 14 are 
+-------+------+---------+
|movieId|userId|   rating|
+-------+------+---------+
|     85|    14|4.0850472|
|     30|    14|2.8075178|
|     59|    14|2.6735032|
|     70|    14|2.4229238|
|     74|    14|2.3021817|
|     87|    14|2.1188734|
|     61|    14|1.9486207|
|     60|    14|1.7559137|
|     92|    14|1.6789422|
|     91|    14| 1.457192|
|     32|    14|1.4171797|
|     81|    14|1.2478681|
|     26|    14|1.1856362|
|     46|    14|1.1693587|
|     33|    14|1.1266334|
+-------+------+---------+
only showing top 15 rows

